server/src/controllers/federated_controller.ts

```diff
@@ -0,0 +1,246 @@
+   * @param ws the websocket connection through which the participant and the server communicate
+   */
+  handle(ws: WebSocket): void {
+    if (this.#aggregator === undefined)
+      throw new Error(`no aggregator for task ${this.task.id}`)
+
+    // Try generating a new Client id until there no collision with existing ones
+    let clientId = randomUUID()
+    while (!this.#aggregator.registerNode(clientId)) {
+      clientId = randomUUID()
+    }
+
+    ws.on('message', (data: Buffer) => {
+      const msg: unknown = msgpack.decode(data)
+      if (!FederatedMessages.isMessageFederated(msg)) {
+        debug("invalid federated message received on WebSocket: %o", msg);
+        return // TODO send back error
+      }
+
+      /* 
+      * A new participant joins the task 
+      */
+      if (msg.type === MessageTypes.ClientConnected) {
+        debug(`client ${clientId} joined ${this.task.id}`)
+        this.#participants.set(clientId, ws) // add the new client
+
+        const waitForMoreParticipants = this.#participants.size < MIN_PARTICIPANTS
+        const msg: Messages.AssignNodeID = {
+          type: MessageTypes.AssignNodeID,
+          id: clientId,
+          waitForMoreParticipants
+        }
+        ws.send(msgpack.encode(msg))
+
+        debug("Wait for more participant flag: %o", waitForMoreParticipants)
+        
+        // If we were previously waiting for more participants to join and we now have enough,
+        // broadcast to previously waiting participants that the training can start
+        if (this.#waitingForMoreParticipants && !waitForMoreParticipants) {
+          for (const [participantId, participantWs] of this.#participants.entries()) {
+            if (participantId !== clientId) {
+              debug("Sending enough-participant message to client: %o", participantId)
+              const msg: FederatedMessages.EnoughParticipants = {
+                type: MessageTypes.EnoughParticipants
+              }
+              participantWs.send(msgpack.encode(msg))
+            }
+          }
+        }
+        this.#waitingForMoreParticipants = waitForMoreParticipants // update the attribute
+
+      /* 
+      * A participant leaves the task 
+      */
+      } else if (msg.type === MessageTypes.ClientDisconnected) {
+        debug(`client ${clientId} left ${this.task.id}`)
+
+        this.#participants.delete(clientId)
+        this.#aggregator.removeNode(clientId)
+
+      /* 
+      * A client sends a weight update to the server
+      */
+      } else if (msg.type === MessageTypes.SendPayload) {
+
+        const { payload, round } = msg
+        if (this.#aggregator.isValidContribution(clientId, round)) {
+          // We need to create a promise waiting for the global model before adding the contribution to the aggregator
+          // (so that the aggregation and sending the global model to participants
+          // doesn't happen before the promise is created)
+          this.createPromiseForWeights(ws)
```

Comment:
soo, I'll bother a bit on this, disregard it if you think that it's not the PR to do that. I might also be quite wrong as it's quite uncharted territories for me.

as this is the only place where the aggregator is being added to, we can avoid all this shenanigans with creating and storing `Promise`s. we can have `Aggregator.add` return the result of the aggregation (it works for both fixed and relative number of peers). this way, no need to wait for the result or playing around with Promises and a clearer way to know when the aggregation results are computed.

Reply:
As long as waiting on the `aggregator.add` doesn't block and prevent the server from receiving other messages and contributions, I think this would be a great solution! This PR is getting too big so I'll put this off to the next one